Convert docs to asciidoc (#112)
elastic/ecs-logging-java: docs/setup-log4j2.asciidoc

|===

To include any custom field in the output, use following syntax:

[source,xml]
----
  <EcsLayout>
    <KeyValuePair key="additionalField1" value="constant value"/>
    <KeyValuePair key="additionalField2" value="$${ctx:key}"/>
  </EcsLayout>
----

Custom fields are included in the order they are declared. The values support [lookups](https://logging.apache.org/log4j/2.x/manual/lookups.html).

[float]
=== Structured logging

By leveraging log4j2's `MapMessage` or even by implementing your own `MultiformatMessage` with JSON support,
you can add additional fields to the resulting JSON.

Example:

[source,java]
----
logger.info(new StringMapMessage()
    .with("message", "Hello World!")
    .with("foo", "bar"));
----

If Jackson is on the classpath, you can also use an `ObjectMessage` to add a custom object the resulting JSON.

[source,java]
----
logger.info(new ObjectMessage(myObject));
----

The `myObject` variable refers to a custom object which can be serialized by a Jackson `ObjectMapper`.

Using either will merge the object at the top-level (not nested under `message`) of the log event if it is a JSON object.

> If it's a string, number boolean or an array, it will be converted into a string and added as the `message` property.

Reviewer: ```suggestion
If it's a string, number boolean, or array, it will be converted into a string and added as the `message` property.
```

> The conversion is done in order to avoid mapping conflicts as `message` is typed as a string in the Elasticsearch mapping.

Reviewer: ```suggestion
This conversion avoids mapping conflicts as `message` is typed as a string in the Elasticsearch mapping.
```

[float]
==== Tips

> It's recommended to use existing [ECS fields]({ecs-ref}/ecs-field-reference.html).

Reviewer: ```suggestion
We recommend using existing {ecs-ref}/ecs-field-reference.html[ECS fields].
```

> If there is no appropriate ECS field,
> consider prefixing your fields with `labels.`, as in `labels.foo`, for simple key/value pairs.
> For nested structures consider prefixing with `custom.` to make sure you won't get conflicts if ECS later adds the same fields but with a different mapping.

Reviewer: ```suggestion
If there is no appropriate ECS field,
consider prefixing your fields with `labels.`, as in `labels.foo`, for simple key/value pairs.
For nested structures, consider prefixing with `custom.`. This approach protects against conflicts in case ECS later adds the same fields but with a different mapping.
```
